brightwind/analyse/analyse.py

```diff
@@ -2245,3 +2252,295 @@ def calc_air_density(temperature: Union[float, pd.Series, pd.DataFrame],
+    :type air_pressure_hPa:         float or pandas.Series
+    :param air_temperature_degC:        Air temperature in degrees Celsius.
+    :type air_temperature_degC:         float or pandas.Series
+    :return:                        Air density in kilograms per cubic meter (kg/m^3).
+                                    Output type depends on input type. If all inputs are float,
+                                    output is float. If any input is pandas.Series,
+                                    output is pandas.Series.
+    :rtype:                         float or pandas.Series
+
+        **Calculation method:**
+        The air density is calculated as the sum of the densities of the dry air component 
+        and the water vapour component using the equation:
+
+            air_density = (P_d / (R_d * T)) + (P_v / (R_v * T))
+
+        where:
+            P_d = dry air partial pressure (Pa) = total pressure - vapour pressure  
+            P_v = vapour pressure (Pa)  
+            R_d = specific gas constant for dry air (287.05 J/kg*K)  
+            R_v = specific gas constant for water vapour (461.5 J/kg*K)  
+            T   = air temperature in Kelvin (°C + 273.15)
+
+    **Example usage**
+    ::
+    import brightwind as bw
+
+    vapour_pressure = 12.3   # hPa
+    pressure = 1013.25       # hPa
+    temperature = 15.0       # °C
+
+    air_density = bw.calc_air_density_from_vapour_prs(vapour_pressure, pressure, temperature)
+    print(air_density)
+    # 1.219391176982024
+
+    # calculate air density from air temperature, dew point temperature and air pressure:
+    T_air = 15
+    T_dew_pt = 14
+    P_air = 1000
+    vapour_prs = bw.calc_vapour_pressure_from_dewpoint(T_dew_pt)
+    air_density = bw.calc_air_density_from_vapour_prs(vapour_prs, P_air, T_air)
+    air_density
+    # 1.2016918568274164 
+
+    # calculate air density from air temperature, relative humidity and air pressure:
+    T_air = 15
+    P_air = 1000
+    RH = 50
+    vapour_prs = bw.calc_vapour_pressure_from_relative_humidity(RH, T_air)
+    air_density = bw.calc_air_density_from_vapour_prs(vapour_prs, P_air, T_air)
+    air_density
+    # 1.2050986677333038
+
+    # calculate air density from air temperature, relative humidity and air pressure series:
+    DATA = bw.load_csv(bw.demo_datasets.demo_data)
+    DATA = bw.apply_cleaning(DATA, bw.demo_datasets.demo_cleaning_file)
+
+    vapour_prs = bw.calc_vapour_pressure_from_relative_humidity(DATA.RH2m, DATA.T2m)
+    air_density = bw.calc_air_density_from_vapour_prs(vapour_prs, DATA.P2m, DATA.T2m)
+    air_density.tail(5)
+    # Timestamp
+    # 2017-11-23 10:10:00    1.195418
+    # 2017-11-23 10:20:00    1.196145
+    # 2017-11-23 10:30:00    1.196055
+    # 2017-11-23 10:40:00    1.195684
+    # 2017-11-23 10:50:00    1.197327
+    # dtype: float64
+    """
+    # Check dimensions of temperature, pressure and rel_humidity_percent if not float or int
```

Comment:
probably we need to have a check for  input type as well...it would be nice to have private functions doing the length and type check as there are several functions having very similar code. I leave @stephenholleran decide if he wants this...I guess that can be a future enhancement no needed now